### PR TITLE
swap: generalize jupiter to any token

### PR DESCRIPTION
**Pull request by 0x0ece**
- added end to end test (requires the full instructions, otherwise accounts may change)
- removed destinationTokenAccount, replaced with our own internal transfer (for security)
- replaced system transfer with token transfer, so we can swap anything
- support for token2022
- simplified API

TODO:
- token2022 should work, but not sure about auto-create ATA for token2022. we can either investigate, or just do it on anchor 0.30 that should be easier
- include wsol wrap automagically like we do for investors
- ~add more tests, e.g. mSOL -> wSOL~

### CODE REVIEW

**Review by yurushao — APPROVED** [review]

File: anchor/programs/glam/src/instructions/jupiter.rs [hunk 1]
```diff
@@ -24,35 +25,48 @@ impl anchor_lang::Id for Jupiter {
...
+    pub fund: Box<Account<'info, FundAccount>>,
+    #[account(mut, seeds = [b"treasury".as_ref(), fund.key().as_ref()], bump)]
+    pub treasury: SystemAccount<'info>,
+
+    /// CHECK: no need to create because input ata should exist to swap, and
+    ///        no need to deser because we transfer_checked from input_ata to
+    ///        input_signer_ata
     #[account(mut)]
-    pub manager: Signer<'info>,
+    pub input_ata: UncheckedAccount<'info>,
```
Comment (yurushao) [inline anchor/programs/glam/src/instructions/jupiter.rs:37]: it's probably better name it input_treasury_ata? same for output ata

File: anchor/tests/glam_jupiter.spec.ts [hunk 2]
```diff
@@ -51,11 +52,359 @@ describe("glam_jupiter", () => {
...
+      addressLookupTableAddresses: [
+        "6TSS8bTxDjGitZgtTc7gdVWkzKdpGLb9QDi7kXAQQXXp"
+      ],
+      prioritizationFeeLamports: 0
+    };
+
+    // Pre-checks: the following accounts should not exist
+    const beforeManagerBalance =
+      await glamClient.provider.connection.getBalance(manager);
+    expect(beforeManagerBalance).toEqual(499_999_999_966_178_300);
```
Comment (yurushao) [inline anchor/tests/glam_jupiter.spec.ts:328]: this is reproducible?

File: anchor/src/client/jupiter.ts [hunk 3]
```diff
@@ -198,76 +124,53 @@ export class JupiterClient {
             "quoteParams must be specified when quoteResponse and swapInstruction are not specified."
           );
         }
-        console.log("Fetching quoteResponse with quoteParams:", quoteParams);
+        // console.log("Fetching quoteResponse with quoteParams:", quoteParams);
         quoteResponse = await this.getQuoteResponse(quoteParams);
+        // console.log("quoteResponse", util.inspect(quoteResponse, false, null));
```
[the remark this answers is not in the document]
Comment (0x0ece) [reply anchor/src/client/jupiter.ts:129]: on this topic, we should prob avoid logging errors in tests when we expect an error :)

File: anchor/tests/glam_jupiter.spec.ts (hunk 2 above)
Comment (0x0ece) [reply anchor/tests/glam_jupiter.spec.ts:328]: lol - well, it is, except if we add more tests before. removing it :D

**Pull request merged**